File: reference/moneymapper_app/lib/services/market_data_service.dart
import 'dart:convert';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:intl/intl.dart';
import 'resilience_utils.dart';

class MarketDataService {
  final _supabase = Supabase.instance.client;
  static const String _cacheKeyRates = "metal_rates_cached_v1";
  static const String _cacheKeyLastFetch = "gold_last_fetch_ts_resilient_v1";

  /// Fetches Metal Rates from Supabase with Caching.
  Future<Map<String, dynamic>> getGoldRate({bool forceRefresh = false}) async {
    final prefs = await SharedPreferences.getInstance();
    
    // 1. Cache Check (15 mins for better live feel)
    if (!forceRefresh) {
      final cached = _readCache(prefs);
      if (cached != null) return cached;
    }

    // 2. Fetch from Supabase
    try {
      final res = await _supabase
          .schema('bse_data')
          .from('live_metal_rates')
          .select();

      if (res == null || (res as List).isEmpty) {
        throw Exception("EMPTY_DATA");
      }

      final List<dynamic> data = res;
      Map<String, dynamic> rates = {};
      double gold24kPrice = 0;

      for (var row in data) {
        String purity = row['purity'] ?? '';
        double price = ResilienceUtils.safeDouble(row['price']);
        rates[purity] = price;

        if (purity.contains('24K')) {
          gold24kPrice = price;
        }
      }

      // Backward compatibility for the ticker (24K 10g)
      rates['price'] = gold24kPrice * 10;
      rates['time'] = DateFormat('hh:mm a').format(DateTime.now());
      rates['status'] = 'Live (IBJA)';

      await _updateCache(prefs, rates);
      return rates;
    } catch (e) {
      // 3. Fallback to Stale Cache
      final stale = _readCache(prefs, force: true);
      if (stale != null) {
        stale['status'] = 'Offline (Stale)';
        return stale;
      }
      return {
        'price': 0.0,
        'status': 'Unavailable',
        'time': 'Sync Failed'
      };
    }
  }

  Map<String, dynamic>? _readCache(SharedPreferences prefs, {bool force = false}) {
    final lastFetchStr = prefs.getString(_cacheKeyLastFetch);
    final cachedDataStr = prefs.getString(_cacheKeyRates);

    if (lastFetchStr == null || cachedDataStr == null) return null;

    final lastFetch = DateTime.parse(lastFetchStr);
    final now = DateTime.now();

    if (force || now.difference(lastFetch).inMinutes < 15) {
      return jsonDecode(cachedDataStr);
    }
    return null;
  }

  Future<void> _updateCache(SharedPreferences prefs, Map<String, dynamic> rates) async {
    await prefs.setString(_cacheKeyRates, jsonEncode(rates));
    await prefs.setString(_cacheKeyLastFetch, DateTime.now().toIso8601String());
  }
}

File: src/services/marketDataService.ts
import { supabase } from '../lib/supabase';
import { ResilienceUtils } from './resilienceUtils';

export interface GoldRateData {
  [purity: string]: any;
  price: number;
  time: string;
  status: string;
  updated_at?: string;
}

export class MarketDataService {
  private static readonly CACHE_KEY_RATES = 'metal_rates_cached_v1';
  private static readonly CACHE_KEY_LAST_FETCH = 'gold_last_fetch_ts_resilient_v1';
  private static readonly CACHE_TTL_MS = 15 * 60 * 1000; // 15 minutes

  /**
   * Fetches Metal Rates from Supabase with Caching (15 min TTL).
   * Ported 1:1 from market_data_service.dart.
   */
  async getGoldRate({ forceRefresh = false }: { forceRefresh?: boolean } = {}): Promise<GoldRateData> {
    // 1. Cache Check (15 mins for better live feel)
    if (!forceRefresh) {
      const cached = this.readCache();
      if (cached) return cached;
    }

    // 2. Fetch from Supabase
    try {
      const { data, error } = await supabase
        .schema('bse_data')
        .from('live_metal_rates')
        .select();

      if (error || !data || (Array.isArray(data) && data.length === 0)) {
        throw new Error(error?.message || 'EMPTY_DATA');
      }

      const rates: GoldRateData = {
        price: 0,
        time: '',
        status: 'Live (IBJA)',
      };
      let gold24kPrice = 0;
      let newestUpdatedAt: string | undefined = undefined;

      for (const row of data as any[]) {
        const purity = row.purity ?? '';
        const price = ResilienceUtils.safeDouble(row.price);
        rates[purity] = price;

        if (purity.includes('24K')) {
          gold24kPrice = price;
        }

        if (row.updated_at) {
          if (!newestUpdatedAt || new Date(row.updated_at) > new Date(newestUpdatedAt)) {
            newestUpdatedAt = row.updated_at;
          }
        }
      }

      // Backward compatibility for ticker (24K 10g)
      rates.price = gold24kPrice * 10;
      rates.time = this.formatTime(new Date());
      rates.status = 'Live (IBJA)';
      if (newestUpdatedAt) {
        rates.updated_at = newestUpdatedAt;
      }

      this.updateCache(rates);
      return rates;
    } catch (_e) {
      // 3. Fallback to Stale Cache
      const stale = this.readCache(true);
      if (stale) {
        stale.status = 'Offline (Stale)';
        return stale;
      }
      return {
        price: 0.0,
        status: 'Unavailable',
        time: 'Sync Failed',
      };
    }
  }

  private readCache(force = false): GoldRateData | null {
    try {
      const lastFetchStr = localStorage.getItem(MarketDataService.CACHE_KEY_LAST_FETCH);
      const cachedDataStr = localStorage.getItem(MarketDataService.CACHE_KEY_RATES);

      if (!lastFetchStr || !cachedDataStr) return null;

      const lastFetch = new Date(lastFetchStr).getTime();
      const now = Date.now();

      if (force || now - lastFetch < MarketDataService.CACHE_TTL_MS) {
        return JSON.parse(cachedDataStr);
      }
      return null;
    } catch {
      return null;
    }
  }

  private updateCache(rates: GoldRateData): void {
    try {
      localStorage.setItem(MarketDataService.CACHE_KEY_RATES, JSON.stringify(rates));
      localStorage.setItem(MarketDataService.CACHE_KEY_LAST_FETCH, new Date().toISOString());
    } catch {
      // storage quota or disabled
    }
  }

  private formatTime(date: Date): string {
    let hours = date.getHours();
    const minutes = date.getMinutes();
    const ampm = hours >= 12 ? 'PM' : 'AM';
    hours = hours % 12;
    hours = hours ? hours : 12; // 0 becomes 12
    const minStr = minutes < 10 ? `0${minutes}` : `${minutes}`;
    const hrStr = hours < 10 ? `0${hours}` : `${hours}`;
    return `${hrStr}:${minStr} ${ampm}`;
  }
}

export const marketDataService = new MarketDataService();
